bulldozer/signals.go

```diff
@@ -21,109 +21,389 @@ import (
 	"strings"
 
 	"github.com/palantir/bulldozer/pull"
+	"github.com/pkg/errors"
 	"github.com/rs/zerolog"
 )
 
+type Signal []string
+
 type Signals struct {
-	Labels            []string `yaml:"labels"`
-	CommentSubstrings []string `yaml:"comment_substrings"`
-	Comments          []string `yaml:"comments"`
-	PRBodySubstrings  []string `yaml:"pr_body_substrings"`
-	Branches          []string `yaml:"branches"`
-	BranchPatterns    []string `yaml:"branch_patterns"`
+	Labels            Signal `yaml:"labels"`
+	CommentSubstrings Signal `yaml:"comment_substrings"`
+	Comments          Signal `yaml:"comments"`
+	PRBodySubstrings  Signal `yaml:"pr_body_substrings"`
+	Branches          Signal `yaml:"branches"`
+	BranchPatterns    Signal `yaml:"branch_patterns"`
+	MaxCommits        int `yaml:"max_commits"`
 }
 
-func (s *Signals) Enabled() bool {
+type SignalsMatches struct {
+	Signals
+}
+
+func (s *Signals) size() int {
 	size := 0
 	size += len(s.Labels)
 	size += len(s.CommentSubstrings)
 	size += len(s.Comments)
 	size += len(s.PRBodySubstrings)
 	size += len(s.Branches)
 	size += len(s.BranchPatterns)
-	return size > 0
+	if s.MaxCommits > 0 {
+		size += 1 
+	}
+	return size
 }
 
-// Matches returns true if the pull request meets one or more signals. It also
+func (s *Signals) Enabled() bool {
+	return s.size() > 0
+}
+
+// MatchesAll returns true if the pull request matches ALL of the signals. It also
+// returns a description of the match status. The tag argument appears
+// in this description and indicates the behavior (trigger, ignore) this
+// set of signals is associated with.
+func (s *Signals) MatchesAll(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+	matches := SignalsMatches{}
```

Comment:
It looks like this is an accumulator to matches so you can compare against the input at the end. Does it need to be a full struct or would it also work to use an integer counting all the matches?

Reply:
it doesn't. I added it there in case it was needed for logging or errors later on. It might be helpful if we want to add more details about matches to the UI but that might be premature.